### DISCUSSION TITLE
Shifting gears from Zwift Play buttons

### DISCUSSION THREAD

**Discussion by serbasan** [doudar/SmartSpin2k#595] [Ideas]
Hi Anthony. 

I'm trying out the gear shifting on the Zwift screen and I'm really excited about the feature.
I think I've read that in order to sync SmartSpin2k gear 1 with Zwift gear 1 you have to do calibration. I haven't been able to do this process yet because I can't finish it. You have to reach a speed but even if you do, phase 3 doesn't finish. I'm missing something.

I wanted to ask if there's a possibility that, just like SmartSpin2k sends Zwift what gear it's in, with Zwift Play you can use the gear shift button in SIM mode. I think it might be possible since Zwift would send a command that tells SmartSpin2k to change gear.

Would that be possible? What do you think?

**Comment by doudar** [comment]
I don't know that anybody has documented Zwift's proprietary protocol they use for shifting. What might be possible however is to add support for the Zwift Play Shifters directly to SmartSpin2k, so that you could shift using them.

**Comment by doudar** [comment]
Also, fwiw, you can currently disregard the on screen commands for the calibration. Once the process starts, it's fully automatic on the SmartSpin2k. With your spring loaded return setup however, I'm not sure how it will behave? Since it's using force to end stops, it might trigger early. You could try increasing the value here:
https://github.com/doudar/SmartSpin2k/blob/83ac674679e998acc494d78ca038e52d0d5dac90/src/Main.cpp#L626C4-L626C58
and here:
https://github.com/doudar/SmartSpin2k/blob/83ac674679e998acc494d78ca038e52d0d5dac90/src/Main.cpp#L597

**Comment by serbasan** [comment]
@doudar I've noticed that as of Zwift version 1.78 the gear changes are no longer displayed in the interface.
I've tested my Zoffline server by downgrading to version 1.77 and they appear in the interface again.
It seems to be a Zwift issue. Is this happening to you? What do you think?

**Reply by doudar** [reply]
I haven't had a chance to test it in a couple days - I've been working on an ERG workout screen in the SS2k Config app. 

My guess is that Zwift seems to store device information based on mac address, and so possibly when you reverted your firmware back to a non-supported device, it stopped displaying the gear indications for that device? (even after you upgraded again) 

That or it's a Zwift issue...

**Reply by serbasan** [reply]
It could be what you're saying. The funny thing is that when you downgrade Zwift the gears appear again in the interface. I've tried with several devices and it repeats itself.
When you update your version to 1.78, can you tell me if it still works for you?

**Comment by serbasan** [comment]
<img width="860" alt="Captura de pantalla 2024-11-22 a las 21 08 53" src="https://github.com/user-attachments/assets/000b5879-7d77-4572-ad23-ca1053c75182">
@doudar I have done a clean install of Zwift and now it says that the gears are not compatible with the trainer. This is an intentional change from Zwift 1.78 and should not be a bug. Some code change will be needed to get them back.

"Your smart trainer does not currently support virtual shifting. However, you can use the controls for other actions in Zwift"

**Reply by doudar** [reply]
Thanks for the information!

If you know anyone with a kickr bike 2, or shift (with the newest firmware) let me know. Also any of the newer watt bikes. 

You could also try changing the device name to Wattbike (probably along with a Zwift re-install, because the names seem to be "sticky" on Zwift lately)

**Reply by doudar** [reply]
@cagnulein maybe they switched to looking for device name?

**Comment by cagnulein** [comment]
Could be, didn't try this yet

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno sab 23 nov 2024 alle 01:46 Anthony Doud ***@***.***>
ha scritto:

**Comment by serbasan** [comment]
I've tried. I've changed the name to WattbikeAtom, uninstalled Zwift and deleted the Zwift folder from Documents.

Despite all that I haven't been able to get the new name to appear after installing Zwift. SmartSpin2k appears again.

**Comment by serbasan** [comment]
Hi @doudar ,
What do you think about the gear display in Zwift? Do you think it could be implemented again? It was a great feature while it lasted. Now that I especially need it for races, I’ve been using an iPad in front of the TV with your app as an alternative to see which gear I’m in. I’d really appreciate your thoughts on this. Best regards, and thank you for your work!

**Reply by doudar** [reply]
Def something I want to get implemented. I need some more BLE scans of a Wattbike or a Stages SB20. I've tracked down a SB20, it's just getting time to go there and get the scan of it that I need.

**Reply by serbasan** [reply]
Thank you, @doudar. I hope it won’t be too complicated to achieve. From what I can see, there have been quite a few changes in Zwift’s protocol, especially regarding trainers like the Kickr. It seems Zwift doesn’t make things easy with its protocol, and manufacturers have to constantly adapt to keep their firmware up to date. I’ll be keeping an eye on your upcoming releases and will share the results. Thank you so much for your work.

**Reply by doudar** [reply]
It's been a while. This branch is a work in progress if you want to build it. It is working however with some limitations. Currently you need to connect SmartSpin2k to your network, and then use the Dircon connections for everything except the controller. Pair the controller using BLE. At that point, it's possible to switch back to BLE for all of the other connections, but Zwift will throw a "Not Compatible" warning when you shift. 

Also, this enables shifting using any connected Zwift controllers, not just the SmartSpin2k shifter.

See Pull Request:
https://github.com/doudar/SmartSpin2k/pull/727
